lib/core/src/Cardano/Wallet/CoinSelection/Internal.hs

```diff
@@ -227,30 +227,48 @@ data SelectionParams u = SelectionParams
         :: SelectionStrategy
         -- ^ Specifies which selection strategy to use. See 'SelectionStrategy'.
     }
-    deriving (Eq, Generic, Show)
+    deriving Generic
+
+deriving instance SelectionContext ctx => Eq (SelectionParams ctx)
+deriving instance SelectionContext ctx => Show (SelectionParams ctx)
 
 -- | Indicates that an error occurred while performing a coin selection.
 --
-data SelectionError u
+data SelectionError ctx
     = SelectionBalanceErrorOf
-      (SelectionBalanceError u)
+      (SelectionBalanceError ctx)
     | SelectionCollateralErrorOf
-      (SelectionCollateralError u)
+      (SelectionCollateralError ctx)
     | SelectionOutputErrorOf
-      SelectionOutputError
-    deriving (Eq, Show)
+      (SelectionOutputError ctx)
+
+deriving instance SelectionContext ctx => Eq (SelectionError ctx)
+deriving instance SelectionContext ctx => Show (SelectionError ctx)
+
+-- | Represents an unsuccessful attempt to select collateral.
+--
+data SelectionCollateralError ctx = SelectionCollateralError
+    { largestCombinationAvailable :: Map (UTxO ctx) Coin
+        -- ^ The largest combination of coins available.
+    , minimumSelectionAmount :: Coin
+        -- ^ A lower bound on the sum of coins to be selected as collateral.
+    }
+    deriving Generic
```

Comment:
I've **_forked_** this type to avoid making any changes to the `Internal.Collateral` module, which only needs a single type parameter `u`, which represents unique UTxO identifiers. We **_could_** make the `Collateral` module use `SelectionContext`, but it would add more complication without much benefit.